frontend/src/views/producer/store.js

```diff
@@ -1,63 +1,175 @@
-//     date: '2018-07-12',
-//     time: '16:32',
-//     performer: 'Chris Hemsworth',
-//     contactInformation: '+3589473843',
-//     ticketCatalog: [
-//       {
-//         id: 'ticket-catalog-1',
-//         ticketDescription: 'For children',
-//         price: 10,
-//         availableSeatForThisType: 300,
-//       },
-//     ],
-//     coverImage: 'https://media.giphy.com/media/9vVCPK87Aw6v6/giphy.gif',
-//     themeColor: '#09B0B8',
-//     ageGroupLimit: '13+',
-//     isBilingual: true,
-//   },
-// };
+// @TODO: abstract away fetch call. Make code more legible
+const BASE_URL = 'http://' + window.location.host;
+
+const transformToMap = (arr = []) => {
+  const result = arr.reduce((accumulator, current) => {
+    accumulator[current.id] = current;
+    return accumulator;
+  }, {});
+  return result;
+};
+
+const removeId = item => {
+  delete item.id;
+  return item;
+};
+
+const User = types
+  .model({
+    username: '',
+    token: types.maybe(types.string),
+  })
+  .views(self => ({
+    get isAuthenticated() {
+      return Boolean(self.token);
+    },
+  }));
+
+const UI = types.model({
+  eventList: types.optional(
+    types.model({
+      fetching: false,
+      fetchError: types.optional(
+        types.union(types.string, types.boolean),
+        false,
+      ),
+    }),
+    {},
+  ),
+});
 
 export const RootProducerModel = types
   .model({
     events: types.optional(types.map(EventModel), {}),
     selectedEvent: types.maybe(types.reference(EventModel)),
+    user: types.optional(User, User.create().toJSON()),
+    ui: types.optional(UI, {}),
   })
   .actions(self => ({
-    addEvent: event => self.events.set(event.id, EventModel.create(event)),
+    // hooks
+    afterCreate() {
+      if (window.localStorage.getItem('store'))
+        applySnapshot(self, JSON.parse(window.localStorage.getItem('store')));
+      if (self.events[self.selectedEvent]) self.selectedEvent = null;
+      self.fetchEvents();
+    },
+    // event actions
     selectEvent: id => (self.selectedEvent = id),
-    patchEvent: event => {
-      self.events.set(event.id, event);
+    fetchEvents: flow(function*() {
+      try {
+        self.ui.eventList.fetching = true;
+        const response = yield window.fetch(BASE_URL + '/events');
+        if (!response.ok)
+          throw new Error('Failed to fetch. Server threw HTTP error code');
+        const json = yield response.json();
+        const processedJson = pipeable(json).pipe(
+          parseTo('camelCase'),
+          transformToMap,
+        );
+        self.ui.eventList.fetching = false;
+        applySnapshot(self.events, processedJson);
+        self.selectedEvent = null;
+      } catch (error) {
+        console.error(error);
+        self.ui.eventList.fetching = false;
+        self.ui.eventList.fetchError = 'Could not fetch events ';
+      }
+    }),
+    patchEvent: flow(function*(event) {
+      if (!resolveIdentifier(EventModel, self.events, event.id)) {
```

Comment:
Maybe you could extract this line into a function called `isNewRecord` and call it here to make the code more readable..

Reply:
i did put a todo at the beginning of the file to clean it up later.

Reply:
👍